ports/rp2/machine_adc.c

```diff
@@ -72,18 +82,26 @@ STATIC mp_obj_t machine_adc_make_new(const mp_obj_type_t *type, size_t n_args, s
 
     } else {
         // Get GPIO and check it has ADC capabilities.
-        channel = mp_hal_get_pin_obj(source);
-        if (!ADC_IS_VALID_GPIO(channel)) {
+        const machine_pin_obj_t *pin = machine_pin_find(source);
+        channel = pin->id;
```

Comment:
If pin is external, is `pin->id` an analog channel number, or is it the external pin number?  Where is this number set in the pin object?

Reply:
It's an external pin number, auto-generated from 0-> `MICROPY_HW_PIN_EXT_COUNT`.

Reply:
OK, so it shouldn't really be called `channel` then, it's a pin number.

As I mentioned above, I think the pin object itself should be passed to the `machine_pin_ext_is_adc()` function, and then maybe that function can map it to an actual Nina ADC channel.

Reply:
Ah never mind, there's actually a GPIO29.

Reply:
> I think the pin object itself should be passed to the machine_pin_ext_is_adc() function

This is done. Do you also want to rename the `channel` variable in `machine_adc.c` ?

Reply:
Yeah, maybe call it `adc_source`?

Reply:
Not so sure about this, so it's eventually converted to a `channel` whether it's a pin or a channel already, seems for consistency maybe add a `machine_pin_ext_to_chan(x)` ?

Reply:
For example:
```C
        machine_pin_ext_config(channel, MACHINE_PIN_MODE_ANALOG, 0);
        channel = machine_pin_ext_adc_from_gpio(channel);
```

Then have a `machine_pin_ext_read_u16(uint32_t channel)` instead ?

Reply:
Yes, but it probably should be that `machine_pin_ext_config()` and `machine_pin_ext_adc_from_gpio()` both take a pin object, not a channel integer, as their first argument.

Reply:
Yes that's possible, but to configure the pin in the same if/else after checking for valid pin, there will be an extra `machine_pin_find(source);` is that okay ?

Reply:
Can't you just reuse the `pin` object that's already been found?

Reply:
Like this:
```c
    } else if (pin && ADC_IS_VALID_GPIO(pin)) {
        // Configure the GPIO pin in ADC mode.
        adc_gpio_init(pin->id);
        channel = ADC_CHANNEL_FROM_GPIO(pin);
```

Reply:
> there will be an extra machine_pin_find(source)

machine_pin_find() already exists in the rp2 port.